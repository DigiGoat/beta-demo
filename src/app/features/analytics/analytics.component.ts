import { Component, ElementRef, type OnInit } from '@angular/core';
import { ConfigService } from '../../services/config/config.service';
import { PlatformService } from '../../services/platform/platform.service';

@Component({
  selector: 'app-analytics',
  templateUrl: './analytics.component.html',
  styleUrl: './analytics.component.scss'
})
export class AnalyticsComponent implements OnInit {
  constructor(private config: ConfigService, private platformService: PlatformService, private el: ElementRef<HTMLElement>) { }
  ngOnInit() {
    if (this.platformService.isDev) {
      this.el.nativeElement.innerHTML = '<!-- Analytics disabled in development mode -->';
    } else if (this.platformService.isServer) {
      if (this.config.analytics.clarity) {
        this.el.nativeElement.innerHTML += `<script>(function (c, l, a, r, i, t, y) { c[a] = c[a] || function () { (c[a].q = c[a].q || []).push(arguments); }; t = l.createElement(r); t.async = 1; t.src = "https://www.clarity.ms/tag/" + i; y = l.getElementsByTagName(r)[0]; y.parentNode.insertBefore(t, y); })(window, document, "clarity", "script", "${this.config.analytics.clarity}");</script>`;
      }
      if (this.config.analytics.gtag) {
        this.el.nativeElement.innerHTML += `<script src="https://www.googletagmanager.com/gtag/js?id=${this.config.analytics.gtag}" async></script>`;
        this.el.nativeElement.innerHTML += '<script>window.dataLayer = window.dataLayer || []; function gtag() { dataLayer.push(arguments); }</script>';
      }
    } else if (this.platformService.isBrowser) {
      const color = window.matchMedia('(prefers-color-scheme: dark)') ? 'Dark' : 'Light';
      if ('clarity' in window) {
        window.clarity('set', 'Color Scheme', color);
      }
      if ('gtag' in window) {
        window.gtag('set', 'Color Scheme', color);
        window.gtag('js', new Date());
        window.gtag('config', this.config.analytics.gtag!, { send_page_view: false });
      }
    }
  }
}
