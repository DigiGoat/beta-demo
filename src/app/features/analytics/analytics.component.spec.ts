import { ComponentFixture, TestBed } from '@angular/core/testing';

import { ConfigServiceMock, EmptyConfigServiceMock } from '../../services/config/config.service.mock';
import type { PlatformService } from '../../services/platform/platform.service';
import { AnalyticsComponent } from './analytics.component';
const clarity = jest.fn();
const gtag = jest.fn();
describe('AnalyticsComponent', () => {
  let component: AnalyticsComponent;
  let fixture: ComponentFixture<AnalyticsComponent>;
  let html: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [AnalyticsComponent]
    })
      .compileComponents();

    fixture = TestBed.createComponent(AnalyticsComponent);
    component = fixture.componentInstance;
    html = fixture.nativeElement;
  });

  it('should create', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
  });
  describe('With a config', () => {
    beforeEach(() => {
      component['config'] = ConfigServiceMock;
    });
    describe('In Dev Mode', () => {
      beforeEach(() => {
        component['platformService'] = { isServer: false, isBrowser: true, isDev: true } as PlatformService;
        fixture.detectChanges();
      });
      it('should disable analytics', () => {
        expect(html.innerHTML).toContain('Analytics disabled in development mode');
      });
      it('should match the snapshot', () => {
        expect(fixture).toMatchSnapshot();
      });
    });
    describe('On the server', () => {
      beforeEach(() => {
        component['platformService'] = { isServer: true, isBrowser: false, isDev: false } as PlatformService;
        fixture.detectChanges();
      });
      it('should add Clarity script', () => {
        const elements = html.querySelectorAll('script');
        const script = Array.from(elements).filter((el) => el.innerHTML.includes('clarity'));
        expect(script).toHaveLength(1);
      });
      it('should add gtag scripts', () => {
        const elements = html.querySelectorAll('script');
        const script = Array.from(elements).filter((el) => el.innerHTML.includes('gtag') || el.src.includes('gtag'));
        expect(script).toHaveLength(2);
      });
      it('should match the snapshot', () => {
        expect(fixture).toMatchSnapshot();
      });
    });
    describe('In the Browser', () => {
      beforeEach(() => {
        component['platformService'] = { isServer: false, isBrowser: true, isDev: false } as PlatformService;
        (window as unknown as { clarity: object; }).clarity = clarity;
        (window as unknown as { gtag: object; }).gtag = gtag;
      });
      it('should configure clarity', () => {
        fixture.detectChanges();
        expect(clarity).toHaveBeenCalledTimes(1);
        expect(clarity).toHaveBeenCalledWith('set', 'Color Scheme', 'Dark');
      });
      it('should configure gtag', () => {
        jest.spyOn(window, 'matchMedia').mockImplementation(() => false as unknown as MediaQueryList);
        fixture.detectChanges();
        expect(gtag).toHaveBeenCalledTimes(3);
        expect(gtag).toHaveBeenNthCalledWith(1, 'set', 'Color Scheme', 'Light');
        expect(gtag).toHaveBeenNthCalledWith(2, 'js', expect.any(Date));
        expect(gtag).toHaveBeenNthCalledWith(3, 'config', 'TEST_GTAG', { send_page_view: false });
      });
      it('should match the snapshot', () => {
        expect(fixture).toMatchSnapshot();
      });
    });
  });
  describe('Without a config', () => {
    beforeEach(() => {
      component['config'] = EmptyConfigServiceMock;
      fixture.detectChanges();
      html = fixture.nativeElement;
    });
    describe('On the server', () => {
      beforeEach(() => {
        component['platformService'] = { isServer: true, isBrowser: false, isDev: false } as PlatformService;
        fixture.detectChanges();
      });
      it('should not add scripts', () => {
        const elements = html.querySelectorAll('script');
        expect(elements).toHaveLength(0);
      });
      it('should match the snapshot', () => {
        expect(fixture).toMatchSnapshot();
      });
    });
    describe('In the Browser', () => {
      beforeEach(() => {
        component['platformService'] = { isServer: false, isBrowser: true, isDev: false } as PlatformService;
        (window as unknown as { clarity: object; }).clarity = clarity;
        (window as unknown as { gtag: object; }).gtag = gtag;
        fixture.detectChanges();
      });
      it('should not configure clarity', () => {
        expect(clarity).not.toHaveBeenCalled();
      });
      it('should configure gtag', () => {
        expect(gtag).not.toHaveBeenCalled();
      });
      it('should match the snapshot', () => {
        expect(fixture).toMatchSnapshot();
      });
    });
  });
});
