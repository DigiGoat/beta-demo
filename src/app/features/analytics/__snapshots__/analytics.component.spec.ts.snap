// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnalyticsComponent With a config In Dev Mode should match the snapshot 1`] = `
<app-analytics
  config={[Function Object]}
  el={[Function _ElementRef]}
  platformService={[Function Object]}
>
</app-analytics>
`;

exports[`AnalyticsComponent With a config In the Browser should match the snapshot 1`] = `
<app-analytics
  config={[Function Object]}
  el={[Function _ElementRef]}
  platformService={[Function Object]}
>
</app-analytics>
`;

exports[`AnalyticsComponent With a config On the server should match the snapshot 1`] = `
<app-analytics
  config={[Function Object]}
  el={[Function _ElementRef]}
  platformService={[Function Object]}
>
  <script>
    (function (c, l, a, r, i, t, y) { c[a] = c[a] || function () { (c[a].q = c[a].q || []).push(arguments); }; t = l.createElement(r); t.async = 1; t.src = "https://www.clarity.ms/tag/" + i; y = l.getElementsByTagName(r)[0]; y.parentNode.insertBefore(t, y); })(window, document, "clarity", "script", "TEST_CLARITY");
  </script><script
    async=""
    src="https://www.googletagmanager.com/gtag/js?id=TEST_GTAG"
  /><script>
    window.dataLayer = window.dataLayer || []; function gtag() { dataLayer.push(arguments); }
  </script>
</app-analytics>
`;

exports[`AnalyticsComponent Without a config In the Browser should match the snapshot 1`] = `
<app-analytics
  config={[Function Object]}
  el={[Function _ElementRef]}
  platformService={[Function Object]}
>
</app-analytics>
`;

exports[`AnalyticsComponent Without a config On the server should match the snapshot 1`] = `
<app-analytics
  config={[Function Object]}
  el={[Function _ElementRef]}
  platformService={[Function Object]}
>
</app-analytics>
`;
